File: src/aoc2015/Day05.java
package aoc2015;

import java.io.IOException;
import java.nio.file.*;
import java.util.List;
import java.util.regex.Pattern;

class Day05 {

  private static int partOne(List<String> list) {
    int count = 0;
    for (String s : list) {
      if (!Pattern.matches("(.*[aeiou]){3}.*", s))
        continue;
      if (!Pattern.matches(".*(.)\\1.*", s))
        continue;
      if (Pattern.matches(".*(ab|cd|pq|xy).*", s))
        continue;
      count++;
    }
    return count;
  }

  private static int partTwo(List<String> list) {
    int count = 0;
    for (String s : list) {
      if (!Pattern.matches(".*(..).*\\1.*", s))
        continue;
      if (!Pattern.matches(".*(.)\\w\\1.*", s))
        continue;
      count++;
    }
    return count;
  }

  public static void main(String[] args) throws IOException {
    List<String> s = Files.readAllLines(Paths.get("./input/2015/Day05_input.txt"));
    System.out.println("Part One = " + partOne(s));
    System.out.println("Part Two = " + partTwo(s));
  }
}

File: src/aoc2017/kot/Day05.kt
package aoc2017.kot

import java.io.File

object Day05 {

  fun partOne(input: List<Int>): Int = solve(input) { _ -> 1 }

  fun partTwo(input: List<Int>): Int = solve(input) { it -> if (it >= 3) -1 else 1 }

  fun solve(input: List<Int>, change: (Int) -> Int): Int {
    val mem = input.toMutableList()
    var pos = 0
    var count = 0
    while (pos < mem.size) {
      val off = mem[pos]
      mem[pos] += change(mem[pos])
      pos += off
      count++
    }
    return count
  }
}

fun main(args: Array<String>) {
  val input = File("./input/2017/Day05_input.txt").readLines().map { it.toInt() }
  println("Part One = ${Day05.partOne(input)}")
  println("Part Two = ${Day05.partTwo(input)}")
}
